// playwright/lib/mcp/browser/tools/video.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var video_exports = {};
__export(video_exports, {
  default: () => video_default
});
module.exports = __toCommonJS(video_exports);
var import_mcpBundle = require("playwright-core/lib/mcpBundle");
var import_tool = require("./tool");
const startVideo = (0, import_tool.defineTabTool)({
  capability: "devtools",
  schema: {
    name: "browser_start_video",
    title: "Start video",
    description: "Start video recording",
    inputSchema: import_mcpBundle.z.object({
      size: import_mcpBundle.z.object({
        width: import_mcpBundle.z.number().describe("Video width"),
        height: import_mcpBundle.z.number().describe("Video height")
      }).optional().describe("Video size")
    }),
    type: "readOnly"
  },
  handle: async (tab, params, response) => {
    await tab.page.video().start({ size: params.size });
    response.addTextResult("Video recording started.");
  }
});
const stopVideo = (0, import_tool.defineTabTool)({
  capability: "devtools",
  schema: {
    name: "browser_stop_video",
    title: "Stop video",
    description: "Stop video recording",
    inputSchema: import_mcpBundle.z.object({
      filename: import_mcpBundle.z.string().optional().describe("Filename to save the video")
    }),
    type: "readOnly"
  },
  handle: async (tab, params, response) => {
    const resolvedFile = await response.resolveFile({ prefix: "video", ext: "webm", suggestedFilename: params.filename }, "Video");
    await tab.page.video().stop({ path: resolvedFile.fileName });
    await response.addFileResult(resolvedFile, null);
  }
});
var video_default = [
  startVideo,
  stopVideo
];
